### PR TITLE
LibJS: Add a QueueingConsoleClient to capture early console commands

### PR DESCRIPTION
**Pull request by davidot**
Now you can `console.log` in HTML to your hearts content

### CODE REVIEW

File: Base/res/html/misc/welcome.html [hunk 1]
```diff
@@ -1,6 +1,11 @@
 <!DOCTYPE html>
 <html lang="en">
 <head>
+    <script>
+        let tempObj = [1, 2, 3];
+        console.log('This is a console.log from within <head> before anything else! What follows is a temporary array:', tempObj);
+        tempObj = null;
+    </script>
```
Comment (davidot) [inline Base/res/html/misc/welcome.html:8]: Maybe this should be tested/shown in a separate page?

File: Userland/Libraries/LibJS/Console.cpp [hunk 2]
```diff
@@ -6,14 +6,110 @@
...
+        String string_data {};
+        Variant<Empty, Vector<String>, MarkedVector<Value>> arguments {};
+    };
+
+    Vector<ConsoleCommand> m_queued_commands;
+};
+
 Console::Console(GlobalObject& global_object)
     : m_global_object(global_object)
+    , m_client(new QueueingConsoleClient(*this))
```
Comment (davidot) [inline Userland/Libraries/LibJS/Console.cpp:112]: This raw `new` (and the `delete` below) irk me a great deal. But unless we want to make ConsoleClient RefCounted I don't know how to nicely solve this.

(I don't know why it is a raw pointer currently but didn't just want to change that)

Comment (davidot) [reply Userland/Libraries/LibJS/Console.cpp:112]: And this immediately caused problems with leaks, maybe we should make it a RefPtr after all...

**Review by linusg — COMMENTED** [review]
Looks ok, but strongly feels like a hack. Why can't we set up the proper console client a bit earlier?

**Comment by davidot** [comment]
> Looks ok, but strongly feels like a hack. Why can't we set up the proper console client a bit earlier?

That could be nicer, but who should own the QueueingConsoleClient?

**Comment by linusg** [comment]
It wouldn't exist?

> This is mostly useful in for example LibWeb where the console client is only created once the page has finished loading. But that meant that console.log calls within the source of the page were never shown anywhere.

Ideally we'd just get rid of this issue.

**Comment by davidot** [comment]
> It wouldn't exist?

> Ideally we'd just get rid of this issue.

I'll have a look why it is this case, I'll put this PR into draft until I've found anything.

**Comment by AtkinsSJ** [comment]
I had the same reaction as Linus. Main issue I could see is we need the Interpreter to exist, and it's made on-demand by the Document. So no Document means no Interpreter. If that logic could be moved out of the document it would be fine though. I don't know anything about how LibJS works to know if that's feasible or correct.

**Pull request closed without merging**